controllers/provider-alicloud/pkg/controller/infrastructure/actuator.go

```diff
@@ -265,8 +283,96 @@ func computeProviderStatusVSwitches(infrastructure *alicloudv1alpha1.Infrastruct
 	return vswitchesToReturn, nil
 }
 
+// getSeedCloudProviderCredentials gets Seed's cloud provider's credentials
+func (a *actuator) getSeedCloudProviderCredentials(ctx context.Context) (*alicloud.Credentials, error) {
+	secretRef := &corev1.SecretReference{
+		Name:      SeedCloudProviderSecretName,
+		Namespace: SeedCloudProviderSecretNamespace,
+	}
+	credentials, err := alicloud.ReadCredentialsFromSecretRef(ctx, a.client, secretRef)
+	if err != nil {
+		return nil, err
+	}
+	return credentials, nil
+}
+
+// shareCustomizedImages checks whether Shoot's Alicloud account has permissions to use the customized images. If it can't
+// access them, these images will be shared with it from Seed's Alicloud account.
+func (a *actuator) shareCustomizedImages(ctx context.Context, infra *extensionsv1alpha1.Infrastructure, cluster *extensioncontroller.Cluster) error {
+	regionID := cluster.Shoot.Spec.Region
+	if a.alicloudECSClient == nil {
+		a.logger.Info("Creating Alicloud ECS client for Seed", "infrastructure", infra.Name)
+		seedCloudProviderCredentials, err := a.getSeedCloudProviderCredentials(ctx)
+		if err != nil {
+			return err
+		}
+		a.alicloudECSClient, err = a.newClientFactory.NewECSClient(ctx, regionID, seedCloudProviderCredentials.AccessKeyID, seedCloudProviderCredentials.AccessKeySecret)
+		if err != nil {
+			return err
+		}
+	}
+
+	workers := cluster.Shoot.Spec.Provider.Workers
+	_, shootCloudProviderCredentials, err := a.getConfigAndCredentialsForInfra(ctx, infra)
+	if err != nil {
+		return err
+	}
+	a.logger.Info("Creating Alicloud ECS client for Shoot", "infrastructure", infra.Name)
+	shootAlicloudECSClient, err := a.newClientFactory.NewECSClient(ctx, regionID, shootCloudProviderCredentials.AccessKeyID, shootCloudProviderCredentials.AccessKeySecret)
+	if err != nil {
+		return err
+	}
+	a.logger.Info("Creating Alicloud STS client for Shoot", "infrastructure", infra.Name)
+	shootAlicloudSTSClient, err := a.newClientFactory.NewSTSClient(ctx, regionID, shootCloudProviderCredentials.AccessKeyID, shootCloudProviderCredentials.AccessKeySecret)
+	if err != nil {
+		return err
+	}
+
+	shootCloudProviderAccountID, err := shootAlicloudSTSClient.GetAccountIDFromCallerIdentity(ctx)
+	if err != nil {
+		return err
+	}
+
+	a.logger.Info("Sharing customized image with Shoot's Alicloud account from Seed", "infrastructure", infra.Name)
+	for _, worker := range workers {
+		imageName := worker.Machine.Image.Name
+		imageVersion := worker.Machine.Image.Version
+		imageID, err := confighelper.FindImageForRegion(a.machineImageMapping, imageName, imageVersion, regionID)
```

Comment:
I see a problem here: Let's assume a shoot with image `foo` and version `1.0`. Now, somebody deploys a new Alicloud extension with a new configmap that no longer contains the `1.0` version. Then, the infrastructure reconciliation for this shoot will always fail.
We have the same problem with the worker controller. Here, we put the real image ID information into the status:

```yaml
...
status:
  providerStatus:
    apiVersion: alicloud.provider.extensions.gardener.cloud/v1alpha1
    kind: WorkerStatus
    machineImages:
    - id: coreos_2023_4_0_64_30G_alibase_20190319.vhd
      name: coreos-alicloud
      version: 2023.4.0
```

I suggest to do either do the same in the infrastructure controller, or do fall back checking the worker status in case the image ID is not found in the config map. See also https://github.com/gardener/gardener-extensions/pull/217 for more info.

Reply:
It is an interesting case. Image sharing is a special step prior to Machine creation. I think we could just `skip sharing` if the image id can't be found according to version and name.  This is because, during whole shoot reconciliation, Machine process loop alway do the same image id searching. 
- If it is found, that means in the infrastructure loop image has already been shared.
- If it is not found, even in the infrasturcture loop the image can be found by some means and shared, it doesn't make any sense. Because machine loop will always fail.

Hope I explained it.

Reply:
Thanks for advice! However, Imho, this is different from the previous worker controller problem.

1. In worker controller, it needs to find a valid imageID to be able to deploy the machine.
1. In infrastructure controller, it only needs to share the image the extension provided to Shoot's alicloud account when the Shoot doesn't have access to the image. Theoretically, if the Shoot has already been reconciled before, the image should have also already been shared with its Alicloud account, thus, no more sharing is needed. In this case, not finding the imageID is okay.

I do, however, see a potential issue: Assume a Shoot with image `foo` and version `1.0`. As new Alicloud extension is deployed, and the new configmap no longer contains the version `1.0` `foo` image. Since we assume the image has already been shared with its Alicloud account, no sharing will occur. But, somehow, the sharing from Seed's Alicloud account to Shoot's Alicloud account is revoked (which normally shouldn't happen), Shoot worker will thus no longer be able to reconcile.

Do we need to address this problem? If we do, then surely we can store the status in infrastructureStatus (or check workerStatus). But if this is not necessary, then we can simply remove `return err` when imageID is not found by `FindImageForRegion()` (yes, in any case, simply returning err here is not proper)

Reply:
> I see a problem here: Let's assume a shoot with image `foo` and version `1.0`. Now, somebody deploys a new Alicloud extension with a new configmap that no longer contains the `1.0` version. Then, the infrastructure reconciliation for this shoot will always fail.
> We have the same problem with the worker controller. Here, we put the real image ID information into the status:
> 
> ```yaml
> ...
> status:
>   providerStatus:
>     apiVersion: alicloud.provider.extensions.gardener.cloud/v1alpha1
>     kind: WorkerStatus
>     machineImages:
>     - id: coreos_2023_4_0_64_30G_alibase_20190319.vhd
>       name: coreos-alicloud
>       version: 2023.4.0
> ```
> 
> I suggest to do either do the same in the infrastructure controller, or do fall back checking the worker status in case the image ID is not found in the config map. See also #217 for more info.

Thanks for advice! However, Imho, this is different from the previous worker controller problem.

1. In worker controller, it needs to find a valid imageID to be able to deploy the machine.
1. In infrastructure controller, it only needs to share the image the extension provided to Shoot's alicloud account when the Shoot doesn't have access to the image. Theoretically, if the Shoot has already been reconciled before, the image should have also already been shared with its Alicloud account, thus, no more sharing is needed. In this case, not finding the imageID is okay.

I do, however, see a potential issue: Assume a Shoot with image `foo` and version `1.0`. As new Alicloud extension is deployed, and the new configmap no longer contains the version `1.0` `foo` image. Since we assume the image has already been shared with its Alicloud account, no sharing will occur. But, somehow, the sharing from Seed's Alicloud account to Shoot's Alicloud account is revoked (which normally shouldn't happen), Shoot worker will thus no longer be able to reconcile.

Do we need to address this problem? If we do, then surely we can store the status in infrastructureStatus (or check workerStatus). But if this is not necessary, then we can simply remove `return err` when imageID is not found by `FindImageForRegion()` (yes, in any case, simply returning err here is not proper)

Reply:
Yes, good points colleagues, thanks. @EmoinLanyu Ideally, you would address this point, yes, but I can also understand that a pragmatic way to start is to follow @jia-jerry's approach - only do the sharing once when the `Infrastructure` is initially created, but then not again. This, however, would not be very much controller-like, where the desired state is always reconciled. If the machine image is revoked from end-user account then manual action is always required, something, that we generally try to avoid when we can. And here we clearly can. But I also understand that this is most likely a rare case. Though, I personally would probably simply store the image versions in the infra status (not looking into the `Worker` object), just to be safe.